backend/windmill-api/src/webhook_util.rs

```diff
@@ -0,0 +1,166 @@
+                                            &db,
+                                        )
+                                        .await else {
+                                            tracing::error!("Webhook Message to send - but cannot get workspace settings! Workspace: {workspace_id}");
+                                            continue;
+                                        };
+                                    cache.insert(workspace_id.clone(), webook_opt, Duration::from_secs(30)).await;
+                                    cache.get(&workspace_id).await.unwrap()
+                                }
+                            };
+                            let webook_opt = url_guard.value();
+                            if let Some(url) = webook_opt {
+                                let timer = WEBHOOK_REQUEST_COUNT.start_timer();
+                                let _ = client.post(url).json(&message).send().await;
+                                timer.stop_and_record();
+                                drop(url_guard);
+                            }
+                        },
+                        None => break,
+                    },
+                    _ = futures::future::poll_fn(|cx| cache_purge_interval.poll_tick(cx)) => {
+                        tracing::trace!("Puring Webhook Cache");
+                        cache.purge(10, 0.50).await;
+                    },
+                }
+            }
+        });
+
+        Self { channel: tx }
+    }
+}
+
+#[derive(Clone)]
+pub struct WebhookUtil {
+    workspace_id: Option<String>,
+    shared: Extension<WebhookShared>,
+}
+
+impl WebhookUtil {
+    pub fn send_message(&self, message: WebhookMessage) {
+        let Some(workspace_id) = &self.workspace_id else {
+            return;
+        };
+        let _ = self.shared.channel.send((workspace_id.clone(), message));
+    }
+}
+
+#[async_trait]
+impl<S> FromRequestParts<S> for WebhookUtil
+where
+    S: Send + Sync,
+{
+    type Rejection = (StatusCode, String);
+
+    async fn from_request_parts(
+        parts: &mut Parts,
+        state: &S,
+    ) -> std::result::Result<Self, Self::Rejection> {
+        let original_uri = OriginalUri::from_request_parts(parts, state)
+            .await
+            .ok()
+            .map(|x| x.0)
+            .unwrap_or_default();
+        let path_vec: Vec<&str> = original_uri.path().split("/").collect();
+        let workspace_id = if path_vec.len() >= 4 && path_vec[0] == "" && path_vec[2] == "w" {
+            Some(path_vec[3].to_owned())
+        } else {
+            None
+        };
+
+        let shared = Extension::<WebhookShared>::from_request_parts(parts, state)
+            .await
+            .map_err(|_| {
+                (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    "Could not aquire shared process while retrieving webhook".to_owned(),
+                )
+            })?;
+
+        Ok(Self { workspace_id, shared })
+    }
```

Comment:
You do not need any of this, just do the impl on WebhookShared directly, all the the endpoints where you use it actually use workspace_id underneath.